### PR TITLE
Remove IterativeInverse class

### LINKED ISSUE #4212: C++ cleanup: Remove deprecated classes
Removes:

 - [x] BlockDiagonalMatrix
 - [ ] <s>FilteredMatrix</s>
 - [x] InverseMatrixRichardson
 - [x] IterativeInverse
 - [x] ProductMatrix
 - [x] ProductSparseMatrix
 - [x] ScaledMatrix
 - [x] SchurMatrix
 - [x] ShiftedMatrix
 - [x] ShiftedMatrixGeneralized
 - [x] TransposeMatrix

### LINKED ISSUE #4567: Make LinearOperator work with distributed::Vector
Currently, `distributed::Vector` are (re)initialized incorrectly with the generic `ReinitHelper` in `linear_operator.h`.

### PR DESCRIPTION
**Pull request by tamiko**
Remove `IterativeInverse` and fix up tests.

Drive-by bugfix: Closes #4567

Closes #4212

### CODE REVIEW

File: doc/news/changes/incompatibilities/20170406MatthiasMaier [hunk 1]
```diff
@@ -1,6 +1,18 @@
...
+  <li>SchurMatrix
+  <li>ShiftedMatrix
+  <li>ShiftedMatrixGeneralized
+  <li>TransposeMatrix
+</ul>
+Use the LinearOperator class instead, see the module on
+@ref LAOperators "linear operators" for more details.
 <br>
-(Matthias Maier, 2017/04/06)
+(Matthias Maier, 2017/04/06 - 2017/XX/XX)
```
Comment (davydden) [inline doc/news/changes/incompatibilities/20170406MatthiasMaier]: Just put today's date?

File: tests/arpack/parpack_advection_diffusion_trilinos.cc [hunk 2]
```diff
@@ -288,15 +289,20 @@ void test ()
     SolverControl solver_control     (dof_handler.n_dofs(), 1e-9,/*log_history*/false,/*log_results*/false);
     SolverControl solver_control_lin (dof_handler.n_dofs(), 1e-10,/*log_history*/false,/*log_results*/false);
```
Comment (davydden) [inline tests/arpack/parpack_advection_diffusion_trilinos.cc]: I think this one is not used

File: tests/arpack/step-36_parpack_mf.cc [hunk 3]
```diff
@@ -121,25 +121,30 @@ void test ()
     SolverControl solver_control     (dof_handler.n_dofs(), 1e-9,/*log_history*/false,/*log_results*/false);
     SolverControl solver_control_lin (dof_handler.n_dofs(), 1e-10,/*log_history*/false,/*log_results*/false);
```
Comment (davydden) [inline tests/arpack/step-36_parpack_mf.cc]: Same here

File: tests/arpack/step-36_parpack_mf.cc [hunk 4]
```diff
@@ -121,25 +121,30 @@ void test ()
...
-    shift_and_invert.solver.select("cg");
-    static ReductionControl inner_control_c(dof_handler.n_dofs(), 0.0, 1.e-13);
-    shift_and_invert.solver.set_control(inner_control_c);
+    const auto shift_and_invert =
+      inverse_operator(linear_operator<VectorType>(laplace),
+                       solver_c,
+                       preconditioner);
+    shift_and_invert.reinit_range_vector = [&vectors](VectorType &v,
+                                                bool omit_zeroing_entries) {
+      v.reinit(vectors.size(), omit_zeroing_entries);
```
Comment (davydden) [inline tests/arpack/step-36_parpack_mf.cc]: That looks wrong, the test runs in parallel, if I am not mistaken

File: doc/news/changes/incompatibilities/20170406MatthiasMaier (hunk 1 above)
Comment (tamiko) [inline doc/news/changes/incompatibilities/20170406MatthiasMaier]: Nah. There is more to follow.

**Comment by tamiko** [comment]
@davydden Thanks for the review. I updated the pull request. Apparently I missed a few things when rushing to the airplane (first an hour late and then 20 minutes early - how shall one plan with that? :-P)

**Comment by bangerth** [comment]
These are all classes that were already deprecated before the last release, right?

You'll have to update step-44, though:
```

FAILED: /usr/bin/clang++   -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_IMPLEMENT_CPP0X=1 -DTBB_USE_DEBUG -Iinclude -I/home/bob/source/include -I/home/bob/source/bundled/tbb41_20130401oss/include -I/home/bob/source/bundled/umfpack/UMFPACK/Include -I/home/bob/source/bundled/umfpack/AMD/Include -I/home/bob/source/bundled/boost-1.62.0/include -I/home/bob/source/bundled/muparser_v2_2_4/include -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/home/bob/libs/parmetis-4.0.3/include -I/home/bob/libs/trilinos-12.4.2-Source/include -I/home/bob/libs/petsc-3.6.4/include -I/home/bob/libs/hdf5-1.8.15-patch1/include -I/home/bob/libs/oce-OCE-0.17/include/oce -I/home/bob/libs/p4est-1.1/FAST/include -I/home/bob/libs/slepc-3.6.0/include    -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -std=c++11 -Wno-parentheses -O0 -ggdb -MMD -MT examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o -MF "examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o.d" -o examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o -c /home/bob/source/examples/step-44/step-44.cc
/home/bob/source/examples/step-44/step-44.cc:70:10: fatal error: 'deal.II/lac/iterative_inverse.h' file not found
```

**Comment by jppelteret** [comment]
`step-44` includes `iterative_inverse.h`? Hmm... undoubtedly a mistake on my part :-/

**Comment by tamiko** [comment]
@bangerth Yes - they all got a prominent "use `LinearOperator` instead" deprecation warning.

@jppelteret I am disappointed I would have expected more from you :-P

**Comment by tamiko** [comment]
@bangerth @jppelteret It was just the header that was left over. @jppelteret already did a great job converting everything to LinearOperator :-) That also explains why I didn't find the include (I grepped for the class name).

Let's see whether the build test succeeds now.

**Comment by jppelteret** [comment]
> I am disappointed I would have expected more from you :-P

You and me both @tamiko. You and me both... I am ashamed! I accidentally added it when doing the conversion to `LinearOperator`s. I would like to blame it on 🍺, but in all honesty I was likely as sober as a judge at the time 😞.

**Comment by tamiko** [comment]
@davydden On second thought - job's done. I have set a final date in the changes entry and will close the tracker bug with this PR.

**Comment by tamiko** [comment]
@jppelteret Blame it on the lack of :beers: - you weren't at the Ballmer peak.

**Comment by bangerth** [comment]
On 07/03/2017 04:06 AM, Matthias Maier wrote:
> @bangerth <https://github.com/bangerth> Yes - they all got a prominent "use 
> |LinearOperator| instead" deprecation warning.

My question was more whether we deprecated the class *before the last 
release*. I just don't remember.

**Comment by tamiko** [comment]
> My question was more whether we deprecated the class *before the last 
> release*. I just don't remember.

Yes.

**Comment by davydden** [comment]
@tamiko step-44 is now missing `#include <deal.II/lac/solver_selector.h>`:
```
FAILED: /usr/bin/clang++   -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_IMPLEMENT_CPP0X=1 -DTBB_USE_DEBUG -Iinclude -I/home/bob/source/include -I/home/bob/source/bundled/tbb41_20130401oss/include -I/home/bob/source/bundled/umfpack/UMFPACK/Include -I/home/bob/source/bundled/umfpack/AMD/Include -I/home/bob/source/bundled/boost-1.62.0/include -I/home/bob/source/bundled/muparser_v2_2_4/include -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/home/bob/libs/parmetis-4.0.3/include -I/home/bob/libs/trilinos-12.4.2-Source/include -I/home/bob/libs/petsc-3.6.4/include -I/home/bob/libs/hdf5-1.8.15-patch1/include -I/home/bob/libs/oce-OCE-0.17/include/oce -I/home/bob/libs/p4est-1.1/FAST/include -I/home/bob/libs/slepc-3.6.0/include    -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -std=c++11 -Wno-parentheses -O0 -ggdb -MMD -MT examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o -MF "examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o.d" -o examples/CMakeFiles/step-44.debug.dir/step-44/step-44.cc.o -c /home/bob/source/examples/step-44/step-44.cc
/home/bob/source/examples/step-44/step-44.cc:3244:13: error: use of undeclared identifier 'SolverSelector'
            SolverSelector< Vector<double> > solver_K_Jp_inv;
            ^
/home/bob/source/examples/step-44/step-44.cc:3244:44: error: expected '(' for function-style cast or type construction
            SolverSelector< Vector<double> > solver_K_Jp_inv;
                            ~~~~~~~~~~~~~~ ^
/home/bob/source/examples/step-44/step-44.cc:3244:46: error: use of undeclared identifier 'solver_K_Jp_inv'
            SolverSelector< Vector<double> > solver_K_Jp_inv;
                                             ^
/home/bob/source/examples/step-44/step-44.cc:3245:13: error: use of undeclared identifier 'solver_K_Jp_inv'
            solver_K_Jp_inv.select("cg");
            ^
/home/bob/source/examples/step-44/step-44.cc:3246:13: error: use of undeclared identifier 'solver_K_Jp_inv'
            solver_K_Jp_inv.set_control(solver_control_K_Jp_inv);
            ^
/home/bob/source/examples/step-44/step-44.cc:3248:52: error: use of undeclared identifier 'solver_K_Jp_inv'
                                                   solver_K_Jp_inv,
                                                   ^
/home/bob/source/examples/step-44/step-44.cc:3280:13: error: use of undeclared identifier 'SolverSelector'
            SolverSelector< Vector<double> > solver_K_con_inv;
            ^
/home/bob/source/examples/step-44/step-44.cc:3280:44: error: expected '(' for function-style cast or type construction
            SolverSelector< Vector<double> > solver_K_con_inv;
                            ~~~~~~~~~~~~~~ ^
/home/bob/source/examples/step-44/step-44.cc:3280:46: error: use of undeclared identifier 'solver_K_con_inv'
            SolverSelector< Vector<double> > solver_K_con_inv;
                                             ^
/home/bob/source/examples/step-44/step-44.cc:3281:13: error: use of undeclared identifier 'solver_K_con_inv'
            solver_K_con_inv.select("cg");
            ^
/home/bob/source/examples/step-44/step-44.cc:3282:13: error: use of undeclared identifier 'solver_K_con_inv'
            solver_K_con_inv.set_control(solver_control_K_con_inv);
            ^
```

**Comment by jppelteret** [comment]
@tamiko I abused my power and fixed the `step-44` related problem myself ;-)

**Comment by tamiko** [comment]
@jppelteret That's fine. :-)

**Pull request merged**